src/compiler/scala/tools/nsc/symtab/classfile/JavaSignatureWalker.scala

```diff
@@ -0,0 +1,113 @@
+package scala.tools.nsc.symtab.classfile
+
+import scala.reflect.internal.ClassfileConstants._
+
+abstract class JavaSignatureWalker {
+  def visitName(internalName: CharSequence): Unit
+  def raiseError(msg: String): Unit
+
+  def walk(sig: CharSequence): Unit = {
+    def raiseError(msg: String): Unit = {
+      this.raiseError(s"$msg (while parsing $sig)")
+    }
+    val EOF: Char = '\0'
+    var index = 0
+    val end = sig.length
+    def accept(ch: Char) {
+      if (current() != ch)
+        raiseError(s"Invalid signature. Expected $ch, but found ${current()} at position $index of $sig")
+      index += 1
+    }
+    def skipSubName(isDelimiter: Char => Boolean): Unit = {
+      val start = index
+      while (!{val ch = current(); ch == EOF || isDelimiter(ch)}) { index += 1 }
+    }
+    def appendSubName(builder: java.lang.StringBuilder, isDelimiter: Char => Boolean): CharSequence = {
+      val start = index
+      skipSubName(isDelimiter)
+      builder.append(sig, start, index)
+    }
+    def current(): Char = {
+      if (index < end)
+        sig.charAt(index)
+      else EOF
+    }
+
+    def parseMethodOrFieldType(): Unit = {
+      val tag = current(); index += 1
+      tag match {
+        case OBJECT_TAG =>
+          def parseTypeArgs(): Unit = {
+            if (current() == '<') {
+              accept('<')
+              while (current() != '>') {
+                current() match {
+                  case '+' | '-' =>
+                    index += 1
+                    parseMethodOrFieldType()
+                  case  '*' =>
+                    index += 1
+                  case _ =>
+                    parseMethodOrFieldType()
+                }
+              }
+              accept('>')
+            }
+          }
+
+          val internalName = new java.lang.StringBuilder
+          appendSubName(internalName, c => c == ';' || c == '<')
+          parseTypeArgs()
+          while (current() == '.') {
+            accept('.')
+            internalName.append('$')
+            appendSubName(internalName, c => c == ';' || c == '<' || c == '.')
```

Comment:
We should call `visitName` for intermediate names here. 

```
class A<T> {
  class B<U> {
    class C { }
  }
}

class T {
  A<String>.B<Object>.C foo() { return null; }
}
```

should give two entries in A: `A$B` and `A$B$C`.

No need to update this PR, I'm creating a best blend of your PR and mine (#5925).